PWGHF/TableProducer/treeCreatorOmegacSt.cxx

```diff
@@ -277,15 +290,47 @@ struct HfTreeCreatorOmegacSt {
       const auto& v0TrackPos = v0.posTrack_as<TracksExt>();
       const auto& v0TrackNeg = v0.negTrack_as<TracksExt>();
 
-      std::array<double, 3> m{o2::analysis::pdg::MassProton, o2::analysis::pdg::MassPiMinus, o2::analysis::pdg::MassKMinus};
-      std::array<std::array<float, 3>, 3> p;
-      // TODO: treat properly as omega +-
-      p[0] = {v0TrackPos.px(), v0TrackPos.py(), v0TrackPos.pz()};
-      p[1] = {v0TrackNeg.px(), v0TrackNeg.py(), v0TrackNeg.pz()};
-      p[2] = {bachelor.px(), bachelor.py(), bachelor.pz()};
-      double massOmega1 = RecoDecay::m(p, m);
-      std::swap(p[0], p[1]);
-      double massOmega2 = RecoDecay::m(p, m);
+      if (!v0TrackPos.hasTPC() || !v0TrackNeg.hasTPC() || !bachelor.hasTPC() ||
+          v0TrackPos.tpcNClsFindable() < minNoClsTrackedCascade ||
+          v0TrackNeg.tpcNClsFindable() < minNoClsTrackedCascade ||
+          bachelor.tpcNClsFindable() < minNoClsTrackedCascade) {
+        continue;
+      }
+
+      const auto& v0TrackPr = trackCasc.sign() > 0 ? v0TrackPos : v0TrackNeg;
+      const auto& v0TrackPi = trackCasc.sign() > 0 ? v0TrackNeg : v0TrackPos;
+
+      // track propagation
+      o2::track::TrackParCov trackParCovV0;
+      o2::track::TrackPar trackParV0;
+      o2::track::TrackPar trackParBachelor;
+      std::array<std::array<float, 3>, 2> momentaOmegaDaughters;
+      float cpaOmega = -1;
+      float decayLengthOmega = -1.;
+      float decayLengthOmegaXY = -1.;
+      if (df2.process(getTrackParCov(v0TrackNeg), getTrackParCov(v0TrackPos))) {
+        trackParCovV0 = df2.createParentTrackParCov(0);
+        if (df2.process(trackParCovV0, getTrackParCov(bachelor))) {
+          const auto& secondaryVertex = df2.getPCACandidate();
+          decayLengthOmega = RecoDecay::distance(secondaryVertex, primaryVertexPos);
+          decayLengthOmegaXY = RecoDecay::distanceXY(secondaryVertex, primaryVertexPos);
+          trackParV0 = df2.getTrackParamAtPCA(0);
+          trackParBachelor = df2.getTrackParamAtPCA(1);
+          trackParV0.getPxPyPzGlo(momentaOmegaDaughters[0]);
+          trackParBachelor.getPxPyPzGlo(momentaOmegaDaughters[1]);
+          std::array<float, 3> pVec;
+          df2.createParentTrackParCov().getPxPyPzGlo(pVec);
+          std::array<float, 3> pvPos = {primaryVertex.getX(), primaryVertex.getY(), primaryVertex.getZ()};
+          cpaOmega = RecoDecay::cpa(pvPos, df2.getPCACandidate(), pVec);
```

Comment:
Can you not reuse `primaryVertexPos` here instead of creating `pvPos`? Or is there a difference because of `float` vs `double`?

Reply:
Indeed, I can use `primaryVertexPos`.